### PR TITLE
WriteObject() uses resumable uploads if requested.

### PR DESCRIPTION
**Pull request by coryan**
With this change the application can select resumable uploads for
streams created via WriteObject().

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/google-cloud-cpp/1597)

### CODE REVIEW

File: google/cloud/storage/object_stream.h [hunk 1]
```diff
@@ -123,6 +123,43 @@ class ObjectWriteStream : public std::basic_ostream<char> {
...
+   * the GCS.
+   */
+  std::uint64_t next_expected_byte() const {
+    return buf_->next_expected_byte();
+  }
+
+  /**
+   * Suspends an upload.
+   *
+   * For resumable uploads this function suspends the current upload, future
```
Comment (houglum) [inline google/cloud/storage/object_stream.h]: Looks like this comment didn't get finished

File: google/cloud/storage/examples/storage_object_samples.cc [hunk 2]
```diff
@@ -322,6 +322,72 @@ void WriteLargeObject(google::cloud::storage::Client client, int& argc,
...
+  auto object_name = ConsumeArg(argc, argv);
+  auto session_id = ConsumeArg(argc, argv);
+
+  //! [resume resumable upload]
+  namespace gcs = google::cloud::storage;
+  [](gcs::Client client, std::string bucket_name, std::string object_name,
+     std::string session_id) {
+    gcs::ObjectWriteStream stream =
+        client.WriteObject(bucket_name, object_name,
+                           gcs::RestoreResumableUploadSession(session_id));
```
Comment (frankyn) [inline google/cloud/storage/examples/storage_object_samples.cc]: woah, this is awesome! A user was asking for this in the Go client library.

Comment (coryan) [reply google/cloud/storage/examples/storage_object_samples.cc]: Thanks, does this mean you are Okay with the changes?

Comment (frankyn) [reply google/cloud/storage/examples/storage_object_samples.cc]: Ah.... apologies, I didn't click update for my comment.

Question, how does a user determine the offset when resuming the upload session? Does the library help query the service?

Comment (coryan) [reply google/cloud/storage/examples/storage_object_samples.cc]: Yes, the library automatically queries the service to find out the next expected byte.

Comment (frankyn) [reply google/cloud/storage/examples/storage_object_samples.cc]: Thanks: Request: add a comment stating that the byte offset is queried when a resumable session id is provided.

Other than that LGTM. I appreciate your patience!

**Pull request merged**